app/scripts/background.js

```diff
@@ -94,16 +95,47 @@ const WORKER_KEEP_ALIVE_MESSAGE = 'WORKER_KEEP_ALIVE_MESSAGE';
 /**
  * In case of MV3 we attach a "onConnect" event listener as soon as the application is initialised.
  * Reason is that in case of MV3 a delay in doing this was resulting in missing first connect event after service worker is re-activated.
+ *
+ * @param remotePort
  */
-
 const initApp = async (remotePort) => {
   browser.runtime.onConnect.removeListener(initApp);
   await initialize(remotePort);
   log.info('MetaMask initialization complete.');
 };
 
+/**
+ * Sends a message to the dapp(s) content script to signal it can connect to MetaMask background as
+ * the backend is not active. It is required to re-connect dapps after service worker re-activates.
+ * For non-dapp pages, the message will be sent and ignored.
+ */
+const sendReadyMessageToTabs = async () => {
+  const tabs = await browser.tabs
+    .query({
+      // Query tabs with a url. New tabs and DevTool tabs have no url value. e.g. chrome://extensions, chrome://inspect
+      url: '<all_urls>',
+      windowType: 'normal',
+    })
+    .then((result) => {
+      checkForErrorAndLog();
+      return result;
+    });
+
+  /* TODO we should only sendMessage to dapp tabs, not all tabs. */
+  for (const tab of tabs) {
+    browser.tabs
+      .sendMessage(tab.id, {
+        name: EXTENSION_MESSAGES.READY,
+      })
+      .then(() => {
+        checkForErrorAndLog();
+      });
+  }
+};
+
 if (isManifestV3) {
   browser.runtime.onConnect.addListener(initApp);
+  sendReadyMessageToTabs();
```

Comment:
Should this wait until after initialization? I am not sure if it's possible for this to result in dapps trying to connect too early before initialization has begun, resulting in dropped connection attempts

Reply:
We do not need backend to be complete initialised for DAPPs to connect, onConnect listener is required by the DAPPS to connect which is added at line 137.

Reply:
In a separate task may be we can work to improve the current initialisation logic. It definitely has scope for improvements.

Reply:
It took me a bit to wrap my head around this too. Essentially the `sendMessage` from this function calls `connect` in `contentscript` and/or`ui.js`,  which then triggers the `onConnect` handler above this line. Then we initialize the app. Thanks for helping me understand this better too earlier @jpuri! 🙏

@Gudahtt